packages/muelu/research/graham/mf_example_01.cpp

```diff
@@ -0,0 +1,642 @@
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact
+//                    Jonathan Hu       (jhu@sandia.gov)
+//                    Andrey Prokopenko (aprokop@sandia.gov)
+//                    Ray Tuminaro      (rstumin@sandia.gov)
+//
+// ***********************************************************************
+//
+// @HEADER
+
+// Matrix-free example 01: Xpetra operator that generates 
+// a simple tridiagonal finite difference poisson matrix
+//
+// This solves the problem f''(x) = h^2*(-pi^2)*sin(pi*x)
+// on the domain [0,1] with a finite difference method.
+// This supplies n global DOFs, and breaks it into
+// contiguous segments depending on the number of MPI ranks.
+
+
+// STL includes
+#include <iostream>
+#include <vector>
+#include <set>
+#include <stdio.h>
+#include <random>
+
+// Teuchos includes
+#include "Teuchos_GlobalMPISession.hpp"
+#include "Teuchos_RCP.hpp"
+#include "Teuchos_ParameterList.hpp"
+#include "Teuchos_Time.hpp"
+#include "Teuchos_TimeMonitor.hpp"
+
+// Kokkos include
+#include "Kokkos_Core.hpp"
+//#include "kokkosTools.hpp"
+
+//Tpetra includes
+#include "Tpetra_Map.hpp"
+#include "Tpetra_MultiVector.hpp"
+#include "Tpetra_CrsMatrix.hpp"
+#include "Tpetra_Import.hpp"
+#include "Tpetra_Export.hpp"
+#include "MatrixMarket_Tpetra.hpp"
+
+//Xpetra includes
+#include "Xpetra_TpetraMultiVector.hpp"
+#include "Xpetra_Map.hpp"
+#include "Xpetra_MultiVector.hpp"
+#include "Xpetra_Operator.hpp"
+#include "Xpetra_Import.hpp"
+#include "Xpetra_Export.hpp"
+#include "Xpetra_Utils.hpp"
+
+// Belos includes
+#include "BelosTpetraAdapter.hpp"
+#include "BelosTpetraOperator.hpp"
+#include "BelosBlockGmresSolMgr.hpp"
+#include "BelosBlockCGSolMgr.hpp"
+#include "BelosBiCGStabSolMgr.hpp"
+#include "BelosGCRODRSolMgr.hpp"
+#include "BelosPCPGSolMgr.hpp"
+#include "BelosPseudoBlockCGSolMgr.hpp"
+#include "BelosPseudoBlockGmresSolMgr.hpp"
+#include "BelosPseudoBlockStochasticCGSolMgr.hpp"
+#include "BelosPseudoBlockTFQMRSolMgr.hpp"
+#include "BelosRCGSolMgr.hpp"
+#include "BelosTFQMRSolMgr.hpp"
+
+// MueLu headers
+#include "MueLu_Level.hpp"
+#include "MueLu_Hierarchy.hpp"
+#include "MueLu_HierarchyManager.hpp"
+#include "MueLu_ParameterListInterpreter.hpp"
+#include "MueLu_AmalgamationFactory.hpp"
+#include "MueLu_CoordinatesTransferFactory.hpp"
+#include "MueLu_UncoupledAggregationFactory.hpp"
+#include "MueLu_AggregationExportFactory.hpp"
+#include "MueLu_Factory.hpp"
+
+#include "mpi.h"
```

Comment:
Why is that needed?

Reply:
It probably isn't. This was built externally and included a header with roughly 120 separate include statements across all Trilinos packages, so I copy-pasted the list and then skimmed it and threw out everything that didn't look important.